revocation/internal/crl/crl.go

```diff
@@ -0,0 +1,306 @@
+			return &result.CertRevocationResult{
+				Result:     result.ResultRevoked,
+				CRLResults: crlResults,
+			}
+		}
+	}
+
+	if lastErr != nil {
+		return &result.CertRevocationResult{
+			Result: result.ResultUnknown,
+			CRLResults: []*result.CRLResult{
+				{
+					Result: result.ResultUnknown,
+					URI:    crlURL,
+					Error:  lastErr,
+				}},
+		}
+	}
+
+	return &result.CertRevocationResult{
+		Result:     result.ResultOK,
+		CRLResults: crlResults,
+	}
+}
+
+// Supported checks if the certificate supports CRL.
+func Supported(cert *x509.Certificate) bool {
+	return cert != nil && len(cert.CRLDistributionPoints) > 0
+}
+
+func validate(crl *x509.RevocationList, issuer *x509.Certificate) error {
+	// check signature
+	if err := crl.CheckSignatureFrom(issuer); err != nil {
+		return fmt.Errorf("failed to verify CRL signature: %w", err)
+	}
+
+	// check validity
+	now := time.Now()
+	if !crl.NextUpdate.IsZero() && now.After(crl.NextUpdate) {
+		return fmt.Errorf("expired CRL. Current time %v is after CRL NextUpdate %v", now, crl.NextUpdate)
+	}
+
+	for _, ext := range crl.Extensions {
+		switch {
+		case ext.Id.Equal(oidFreshestCRL):
+			return ErrDeltaCRLNotSupported
+		case ext.Id.Equal(oidIssuingDistributionPoint):
+			// IssuingDistributionPoint is a critical extension that identifies
+			// the scope of the CRL. Since we will check all the CRL
+			// distribution points, it is not necessary to check this extension.
+		default:
+			if ext.Critical {
+				// unsupported critical extensions is not allowed. (See RFC 5280, Section 5.2)
+				return fmt.Errorf("unsupported critical extension found in CRL: %v", ext.Id)
+			}
+		}
+	}
+
+	return nil
+}
+
+// checkRevocation checks if the certificate is revoked or not
+func checkRevocation(cert *x509.Certificate, baseCRL *x509.RevocationList, signingTime time.Time, crlURL string) (*result.CRLResult, error) {
+	if cert == nil {
+		return nil, errors.New("certificate cannot be nil")
+	}
+
+	if baseCRL == nil {
+		return nil, errors.New("baseCRL cannot be nil")
+	}
+
+	// latestTempRevokedEntry contains the most recent revocation entry with
+	// reasons such as CertificateHold or RemoveFromCRL.
+	//
+	// If the certificate is revoked with CertificateHold, it is temporarily
+	// revoked. If the certificate is shown in the CRL with RemoveFromCRL,
+	// it is unrevoked.
+	var latestTempRevokedEntry *x509.RevocationListEntry
+	for i, revocationEntry := range baseCRL.RevokedCertificateEntries {
+		if revocationEntry.SerialNumber.Cmp(cert.SerialNumber) == 0 {
+			extensions, err := parseEntryExtensions(revocationEntry)
+			if err != nil {
+				return nil, err
+			}
+
+			// validate signingTime and invalidityDate
+			if !signingTime.IsZero() && !extensions.invalidityDate.IsZero() &&
+				signingTime.Before(extensions.invalidityDate) {
+				// signing time is before the invalidity date which means the
+				// certificate is not revoked at the time of signing.
+				continue
```

Comment:
OK, `invalidityDate` extension should not be exsit for `RemoveFromCRL`

Reply:
Updated to use `break`

Reply:
I found another case:

For example,
```
               Signing Time
                       |
[--------|-------------|-------------|--------------> Time series
         |                           |
   CertificateHold             KeyCompromise
```
Suppose that `KeyCompromise` appears before `CertificateHold` in the revocation list (the revocation list is not ordered). The `invalidityDate` for `KeyCompromise` is after `signingTime`, so we will `break` here based on your suggestion. However, `CertificateHold` occurred before `signingTime`, which means that the key is invalid between the period of `CertificateHold` and `KeyCompromise`. Therefore, we should return that it was revoked. But if we break here, we will return OK. Is this case possible?

Reply:
I think we need to validate that, a CRL can have multiple entries for same cert? and if it can does  RemoveFromCRL and CertificateHold needs to be in particular order?